crypto/cms/cms_smime.c

```diff
@@ -468,7 +468,8 @@ CMS_ContentInfo *CMS_sign(X509 *signcert, EVP_PKEY *pkey,
         goto err;
 
  merr:
-    CMSerr(CMS_F_CMS_SIGN, ERR_R_MALLOC_FAILURE);
+    if (ERR_peek_error() == 0)
+        CMSerr(CMS_F_CMS_SIGN, ERR_R_MALLOC_FAILURE);
```

Comment:
Wouldn't it be better to push the determination of what kind of error it is down into `CMS_add1_cert()`? Incidentally this could fail to add an error if the error queue is not empty when we enter this function.

Reply:
Hmmm..., Initially I intended to track down all error paths. I managed to do it for `CMS_SignedData_init()`, which is why I changed the `goto merr` to `goto err`.
```
CMS_SignedData_init(cms)
	cms_signed_data_init(cms)
		CMSerr(CMS_F_CMS_SIGNED_DATA_INIT, ERR_R_MALLOC_FAILURE);
```

Then I followed `CMS_add1_cert()'...

```
CMS_add1_cert()
	CMS_add0_cert()
		cms_get0_certificate_choices()
		CMS_add0_CertificateChoices()
```
... but when I reached `CMS_add0_CertificateChoices()` I was finally discouraged and stopped digging further :-/

```C
CMS_CertificateChoices *CMS_add0_CertificateChoices(CMS_ContentInfo *cms)
{
    STACK_OF(CMS_CertificateChoices) **pcerts;
    CMS_CertificateChoices *cch;
    pcerts = cms_get0_certificate_choices(cms);
    if (!pcerts)
        return NULL;
    if (!*pcerts)
        *pcerts = sk_CMS_CertificateChoices_new_null();
    if (!*pcerts)
        return NULL;
    cch = M_ASN1_new_of(CMS_CertificateChoices);
    if (!cch)
        return NULL;
    if (!sk_CMS_CertificateChoices_push(*pcerts, cch)) {
        M_ASN1_free_of(cch, CMS_CertificateChoices);
        return NULL;
    }
    return cch;
}
```

I understand your concerns regarding `ERR_peek_error()` because I had them, too. So I'll probably remove the peek again and try another attempt to differentiate the errors instead. Also, I might replace the wild guess `CMSerr(CMS_F_CMS_SIGN, ERR_R_MALLOC_FAILURE)` by a more unspecific error message.